server/schedule/operator/status.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+// OpStatus represents the status of an Operator.
+type OpStatus = uint32
+
+// Status list
+const (
+	// Status list
+	// Just created. Next status: {RUNNING, CANCELED, EXPIRED}.
+	CREATED OpStatus = iota
+	// Started and not finished. Next status: {SUCCESS, CANCELED, REPLACED, TIMEOUT}.
+	STARTED
+	// Followings are end status, i.e. no next status.
+	SUCCESS  // Finished successfully
+	CANCELED // Canceled due to some reason
+	REPLACED // Replaced by an higher priority operator
+	EXPIRED  // Didn't start to run for too long
+	TIMEOUT  // Running for too long
+	// Status list end
+	statusCount    // Total count of status
+	firstEndStatus = SUCCESS
+)
+
+type transition [statusCount][statusCount]bool
+
+// Valid status transition
+var validTrans transition = transition{
+	CREATED: {
+		STARTED:  true,
+		CANCELED: true,
+		EXPIRED:  true,
+	},
+	STARTED: {
+		SUCCESS:  true,
+		CANCELED: true,
+		REPLACED: true,
+		TIMEOUT:  true,
+	},
+	SUCCESS:  {},
+	CANCELED: {},
+	REPLACED: {},
+	EXPIRED:  {},
+	TIMEOUT:  {},
+}
+
+var statusString [statusCount]string = [statusCount]string{
+	CREATED:  "Created",
+	STARTED:  "Started",
+	SUCCESS:  "Success",
+	CANCELED: "Canceled",
+	REPLACED: "Replaced",
+	EXPIRED:  "Expired",
+	TIMEOUT:  "Timeout",
+}
+
+const invalid pdpb.OperatorStatus = pdpb.OperatorStatus_RUNNING + 1
+
+var pdpbStatus [statusCount]pdpb.OperatorStatus = [statusCount]pdpb.OperatorStatus{
+	// FIXME: use a valid status
+	CREATED:  invalid,
+	STARTED:  pdpb.OperatorStatus_RUNNING,
+	SUCCESS:  pdpb.OperatorStatus_SUCCESS,
+	CANCELED: pdpb.OperatorStatus_CANCEL,
+	REPLACED: pdpb.OperatorStatus_REPLACE,
+	// FIXME: use a better status
+	EXPIRED: pdpb.OperatorStatus_TIMEOUT,
```

Comment:
See comments in line 34,35. `TIMEOUT` indicates the operator has been running for too long. `EXPIRED` indicates the operator has been created but not started for too long.

Reply:
Too much state in state machine is not a good thing.

Reply:
> Too much state in state machine is not a good thing.

Which state do you think we can remove?

Reply:
Expired really need it, do we already meet it?

Reply:
> Expired really need it, do we already meet it?

We think we haven't met it. The motivation of `EXPIRED` is to solve "operator leak". It may occur in the future.